token/program-2022/src/extension.rs

```diff
@@ -476,6 +500,108 @@ mod test {
         solana_program::pubkey::Pubkey,
     };
 
+    fn test_mint_transfer_fee() -> MintTransferFee {
+        MintTransferFee {
+            transfer_fee_config_authority: OptionalNonZeroPubkey::try_from(Some(Pubkey::new(
+                &[10; 32],
+            )))
+            .unwrap(),
+            withheld_withdraw_authority: OptionalNonZeroPubkey::try_from(Some(Pubkey::new(
+                &[11; 32],
+            )))
+            .unwrap(),
+            withheld_amount: PodU64::from(u64::MAX),
+            older_transfer_fee: TransferFee {
+                epoch: PodU64::from(1),
+                maximum_fee: PodU64::from(10),
+                transfer_fee_basis_points: PodU16::from(100),
+            },
+            newer_transfer_fee: TransferFee {
+                epoch: PodU64::from(100),
+                maximum_fee: PodU64::from(5_000),
+                transfer_fee_basis_points: PodU16::from(1),
+            },
+        }
+    }
+
+    const MINT_WITH_EXTENSION: &[u8] = &[
+        // base mint
+        1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+        1, 1, 1, 1, 1, 42, 0, 0, 0, 0, 0, 0, 0, 7, 1, 1, 0, 0, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+        2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, // padding
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // account type
+        1, // extension type
+        3, 0, // length
+        32, 0, // data
+        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+        1, 1,
+    ];
+
+    #[test]
+    fn unpack_opaque_buffer() {
+        let state = StateWithExtensions::<Mint>::unpack(MINT_WITH_EXTENSION).unwrap();
+        assert_eq!(state.base, TEST_MINT);
+        let extension = state.get_extension::<MintCloseAuthority>().unwrap();
```

Comment:
So to be clear, a client only knows which `get_extension` type(s) to use for a particular Mint by trying them?

Reply:
At the moment, yes.  Since extensions are meant to be additive, I was thinking that clients and instruction processors just do something like:
```
if let Ok(extension) = state.get_extension::<SomeExtension>() {
    /// do stuff here
} else {
    /// do some other stuff
}
```
But now that you mention it, the `get_account_data_length` instruction will have to extract all of the extension types contained in the mint in order to provide an answer, which means there's definitely a use case for something like:
```
let extensions: Vec<ExtensionType> = state.get_extensions();
```
which loops through all of the TLV entries and extracts the type.  Seems like we can do that in another PR though -- what do you think?

Reply:
Yes! That's exactly where my head was going. I think that method would be super helpful. Separate PR sounds perfect.
On a related topic, do we ever foresee reworking extensions to support multiples of the same ExtensionType?

Reply:
That's a good question too... maybe there's a different API for that? Something like `state.get_extension_index(2)::<Extension>()`? And then some marker on the extensions so that they can allow multiples.  We'll have to rework the internals for that either way, and I think that our current design would allow for it pretty easily

Reply:
> do we ever foresee reworking extensions to support multiples of the same ExtensionType?

I sorta hope never myself